find/client/http/dhash_client.go

```diff
@@ -70,101 +71,115 @@ func NewDHashClient(dhstoreURL, stiURL string, options ...Option) (*DHashClient,
 	}, nil
 }
 
+// Find launches FindAsync in a separate go routine and assembles the result into FindResponse as if it was a synchronous invocation.
 func (c *DHashClient) Find(ctx context.Context, mh multihash.Multihash) (*model.FindResponse, error) {
-	// query value keys from indexer
+	resChan := make(chan model.ProviderResult)
+	errChan := make(chan error)
+
+	go c.FindAsync(ctx, mh, resChan, errChan)
+
+	mhr := model.MultihashResult{
+		Multihash: mh,
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, errors.New("context cancelled")
+		case err := <-errChan:
+			return nil, err
+		case pr, ok := <-resChan:
+			if !ok {
+				return &model.FindResponse{
+					MultihashResults: []model.MultihashResult{mhr},
+				}, nil
+			}
+			mhr.ProviderResults = append(mhr.ProviderResults, pr)
+		}
+	}
+}
+
+// FindAsync implements double hashed lookup workflow. It submits results as they get decrypted and assembled into resChan. If an error occurs it is sent to errChan.
+// Once the workflow is finished both channels are closed.
+func (c *DHashClient) FindAsync(ctx context.Context, mh multihash.Multihash, resChan chan model.ProviderResult, errChan chan error) {
```

Comment:
Since the output channels are only for output, designate them at write-only:
```suggestion
func (c *DHashClient) FindAsync(ctx context.Context, mh multihash.Multihash, resChan chan<- model.ProviderResult, errChan chan<- error) {
```